### ISSUE TITLE
Base rule extension: [no-restricted-exports] Restrict `type` exports

### ISSUE THREAD

**Issue by RobHannay** [typescript-eslint/typescript-eslint#6945]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+new+base+rule+extension%22) and found none that match my proposal.
- [X] I have searched the [current extension rule list](https://typescript-eslint.io/rules/#extension-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] My proposal specifically checks TypeScript syntax, or it proposes a check that requires type information to be accurate.
- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the base rule

https://eslint.org/docs/latest/rules/no-restricted-exports

### Description

I wish to restrict specific named type exports (in my case disallowing a type called `Props`) from ever being exported.

I _don't_ wish to blanket-ban types from having those names.

That's where this rule should come in handy.

### Fail

```typescript
/* @typescript-eslint/no-restricted-exports: ["error", {
    "restrictedNamedExports": ["Props"]
}] */
export type Props = 'me';
```

### Pass

```typescript
/* @typescript-eslint/no-restricted-exports: ["error", {
    "restrictedNamedExports": ["Props"]
}] */
type Props = 'me'; // Not exported - happy for types to have this name if they're not exported

export type AnythingElse = Props & Something;
```

**Comment by bradzacher** [comment]
https://typescript-eslint.io/linting/troubleshooting#how-can-i-ban-specific-language-feature

This FAQ should be of use to you.

---

I think this request is a bit too niche overall for us to create and maintain an extension rule for. It's rare that people want to restrict specific names, and rarer still that people want to restrict specific type names.

**Comment by Josh-Cena** [comment]
@bradzacher The base rule does do exactly those things, though. Sounds okay to build a thin extension.

**Comment by bradzacher** [comment]
In theory it should be a thin layer - but in reality it ends up with a few hundred lines, not including tests.

Theres a chunk of logic we have to re-implement because we have to manually re-implement the "is restricted" logic.

For example - see the extension for [`no-restricted-imports`](https://github.com/typescript-eslint/typescript-eslint/blob/v6/packages/eslint-plugin/src/rules/no-restricted-imports.ts).

If we could simply hook into the logic and go "just before we report, check to see if it's a type" then you'd be golden - but we can only sit in front of eslint.

Additionally this would require type information so that we can determine if your export is a value or a type.

For example we need types to correctly understand these cases:
```ts
export { Props } from 'module';

// or

import { Props } from 'module';
export { Props };
```

Additionally things get funky with "dual scope" names like classes and enums. Should the rule ban cases like

```ts
export enum Props {}
// or
export class Props {}
```

We'll it depends on the motivation of banning a name. If the motivation is to make intellisense better - then yes it should because intellisense will suggest importing these dual scope names in a type location.

There would be a lot more nuance and work involved in creating this properly so it doesn't false negative.
Which is a large maintenance overhead for a pretty niche usecase!
